operator/internal/deploy/chart_manager.go

```diff
@@ -0,0 +1,222 @@
+	skrClient, err := remote.NewRemoteClient(ctx, kcpClient, client.ObjectKeyFromObject(kyma),
+		kyma.Spec.Sync.Strategy, remoteClientCache)
+	if err != nil {
+		return true, err
+	}
+	skrCfg, err := remote.GetRemoteRestConfig(ctx, kcpClient, client.ObjectKeyFromObject(kyma),
+		kyma.Spec.Sync.Strategy)
+	if err != nil {
+		return true, err
+	}
+	argsVals, err := m.generateHelmChartArgs(ctx, kcpClient)
+	if err != nil {
+		return true, err
+	}
+	// TODO(khlifi411): make sure that validating-webhook-config resource is in sync with the secret configuration
+	skrWatcherInstallInfo := prepareInstallInfo(ctx, m.config.WebhookChartPath, ReleaseName, skrCfg, skrClient, argsVals)
+	err = installOrRemoveChartOnSKR(ctx, skrWatcherInstallInfo, ModeInstall, m.config.EnableWebhookPreInstallCheck)
+	if err != nil {
+		return true, err
+	}
+	kyma.UpdateCondition(v1alpha1.ConditionReasonSKRWebhookIsReady, metav1.ConditionTrue)
+	return false, nil
+}
+
+func (m *EnabledSKRWebhookChartManager) RemoveWebhookChart(ctx context.Context, kyma *v1alpha1.Kyma,
+	syncCtx *remote.KymaSynchronizationContext,
+) error {
+	skrCfg, err := remote.GetRemoteRestConfig(ctx, syncCtx.ControlPlaneClient, client.ObjectKeyFromObject(kyma),
+		kyma.Spec.Sync.Strategy)
+	if err != nil {
+		return err
+	}
+	argsVals, err := m.generateHelmChartArgs(ctx, syncCtx.ControlPlaneClient)
+	if err != nil {
+		return err
+	}
+	skrWatcherInstallInfo := prepareInstallInfo(ctx, m.config.WebhookChartPath, ReleaseName, skrCfg,
+		syncCtx.RuntimeClient, argsVals)
+	return installOrRemoveChartOnSKR(ctx, skrWatcherInstallInfo, ModeUninstall, m.config.EnableWebhookPreInstallCheck)
+}
+
+func (m *EnabledSKRWebhookChartManager) generateHelmChartArgs(ctx context.Context,
+	kcpClient client.Client,
+) (map[string]interface{}, error) {
+	watcherList := &v1alpha1.WatcherList{}
+	if err := kcpClient.List(ctx, watcherList, &client.ListOptions{}); err != nil {
+		return nil, fmt.Errorf("error listing watcher resources: %w", err)
+	}
+	if len(watcherList.Items) == 0 {
+		return nil, ErrFoundZeroWatchers
+	}
+	chartCfg := generateWatchableConfigs(watcherList)
+	bytes, err := k8syaml.Marshal(chartCfg)
+	if err != nil {
+		return nil, err
+	}
+
+	kcpAddr, err := m.resolveKcpAddr(ctx, kcpClient)
+	if err != nil {
+		return nil, err
+	}
+	return map[string]interface{}{
+		"triggerLabel":          time.Now().Format(triggerLabelTimeFormat),
+		"kcpAddr":               kcpAddr,
+		"resourcesLimitsMemory": m.config.SkrWebhookMemoryLimits,
+		"resourcesLimitsCPU":    m.config.SkrWebhookCPULimits,
+		customConfigKey:         string(bytes),
+	}, nil
+}
+
+func (m *EnabledSKRWebhookChartManager) resolveKcpAddr(ctx context.Context, kcpClient client.Client) (string, error) {
+	if m.kcpAddr != "" {
+		return m.kcpAddr, nil
+	}
+	// Get external IP from the ISTIO load balancer external IP
+	loadBalancerService := &corev1.Service{}
+	if err := kcpClient.Get(ctx, client.ObjectKey{Name: IngressServiceName, Namespace: IstioSytemNs},
```

Comment:
This get call is not using the informer cache and is thus extremely slow. You should add a watches on services to handle deletion and error correctly for Watcher CR. In general you are missing a state on Watcher CR if the Virtual Service has been created/updated correctly

Reply:
As discussed. This is a bigger refactoring which will be done [here](https://github.com/kyma-project/lifecycle-manager/issues/302)

Reply:
The Service Lookup is not part of a bigger factory and can be fixed by doing a watch with a label IMO. Please share your opinion @ruanxin @adityabhatia

Reply:
As discussed, I will double check if the service is included in the cache and will add it if it's not 👍

Reply:
I have checked and it is added [here](https://github.com/kyma-project/lifecycle-manager/blob/d8f59928a280ac807b3c8b3b6f17df38c7461252/operator/controllers/util.go#L18-L20)